Misc/NEWS.d/next/Library/2023-03-10-13-21-16.gh-issue-102578.-gujoI.rst

```diff
@@ -0,0 +1,24 @@
+Speed up sanity check for ``__setattr__`` and ``__delattr__`` for frozen
+``dataclass``\es by using ``set``-based name lookup rather than ``tuple``\s.
+
+The sanity check is always performed, even if the ``dataclass`` is inherited.
+For example:
+
+.. code-block:: python
+
+    @dataclass(frozen=True)
+    class FrozenBase:
+        x: int
+        y: int
+        ... # N_FIELDS
+
+    class Foo(FrozenBase):
+        def __init__(self, x, y, somevalue, someothervalue):
+            super().__init__(x, y)
+            self.somevalue = somevalue            # takes O(N_FIELDS)
+            self.someothervalue = someothervalue  # takes O(N_FIELDS) time again
+
+    foo = Foo(1, 2, 3, 4)
+    foo.extravalue = extravalue  # takes O(N_FIELDS) time again
+
+This change makes the sanity check :math:`O(1)` in the number of fields.
```

Comment:
Thank you! I now understand perfectly what this speeds up, and agree that it seems like a good idea. But we can be a lot more concise for this news entry :)

```suggestion
Speed up setting or deleting mutable attributes on non-dataclass subclasses of
frozen dataclasses. Due to the implementation of ``__setattr__`` and
``__delattr__`` for frozen dataclasses, this previously had a time complexity
of ``O(n)``. It now has a time complexity of ``O(1)``.
```